overlord/snapshotstate/backend/backend.go

```diff
@@ -323,3 +324,88 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 
 	return nil
 }
+
+type exportMetadata struct {
+	Format int       `json:"format"`
+	Date   time.Time `json:"date"`
+}
+
+// Export allows exporting a given snapshot set to the given writer.
+//
+// XXX: locking? i.e. avoid being able to remove a snapshot while in use
+func Export(ctx context.Context, setID uint64, w io.Writer) error {
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	// write the archives
+	n := 0
+	err := Iter(ctx, func(reader *Reader) error {
+		if reader.SetID == setID {
+			stat, err := reader.Stat()
+			if err != nil {
+				return err
+			}
+			// should never happen
+			if stat.IsDir() {
+				return fmt.Errorf("unexported directory in snapshot: %v", stat.Name())
+			}
+
+			hdr := &tar.Header{
+				Typeflag: tar.TypeReg,
+				Name:     filepath.Base(reader.Name()),
+				Size:     stat.Size(),
+				Mode:     int64(stat.Mode()),
+				ModTime:  stat.ModTime(),
+			}
+			if err = tw.WriteHeader(hdr); err != nil {
+				return err
+			}
+			// open the file
+			f, err := os.Open(reader.Name())
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+
+			if _, err := io.Copy(tw, f); err != nil {
```

Comment:
What's the memory usage like during this operation. It would be good to add a TODO to have a spread test using `memory-observe-do` to ensure that this will work reliably when working with large archives.

Reply:
io.Copy uses a 32Kb buffer during the copy. So there should be no issues as long as passed in io.Writer is not something silly. It will be a http.ResponseWriter in the real usage which should behave sensibly, i.e. it's not different from streaming a snap in /v2/download or similar. Or am I missing something?

Reply:
I think having a TODO and a spread test should suffice.

Reply:
I've added a check to the `snapshot-basic` spread test.